contracts/reward/RewardsDistributor.sol

```diff
@@ -0,0 +1,176 @@
+import {IERC20Ext, IRewardsDistributor} from '../interfaces/rewardDistribution/IRewardsDistributor.sol';
+
+
+contract RewardsDistributor is IRewardsDistributor, PermissionAdmin, ReentrancyGuard, Utils {
+  using SafeERC20 for IERC20Ext;
+  using SafeMath for uint256;
+
+  struct MerkleData {
+    uint256 cycle;
+    bytes32 root;
+    string contentHash;
+  }
+
+  IPool public treasuryPool;
+  MerkleData private merkleData;
+  // wallet => token => claimedAmount
+  mapping(address => mapping(IERC20Ext => uint256)) public claimedAmounts;
+
+  event TreasuryPoolSet(IPool indexed treasuryPool);
+  event Claimed(
+    uint256 indexed cycle,
+    address indexed user,
+    IERC20Ext[] tokens,
+    uint256[] claimAmounts
+  );
+  event RootUpdated(uint256 indexed cycle, bytes32 root, string contentHash);
+
+  constructor(address admin, IPool _treasuryPool) PermissionAdmin(admin) {
+    treasuryPool = _treasuryPool;
+  }
+
+  receive() external payable {}
+
+  function getMerkleData() external view returns (MerkleData memory) {
+    return merkleData;
+  }
+
+  /**
+   * @dev Checks whether a claim is valid or not
+   * @param cycle cycle number
+   * @param index user reward info index in the array of reward info
+   * during merkle tree generation
+   * @param user wallet address of reward beneficiary
+   * @param tokens array of tokens claimable by reward beneficiary
+   * @param cumulativeAmounts cumulative token amounts claimable by reward beneficiary
+   * @param merkleProof merkle proof of claim
+   * @return true if valid claim, false otherwise
+   **/
+  function isValidClaim(
+    uint256 cycle,
+    uint256 index,
+    address user,
+    IERC20Ext[] calldata tokens,
+    uint256[] calldata cumulativeAmounts,
+    bytes32[] calldata merkleProof
+  ) external view override returns (bool) {
+    if (cycle != merkleData.cycle) return false;
+    bytes32 node = keccak256(abi.encodePacked(cycle, index, user, tokens, cumulativeAmounts));
+    return MerkleProof.verify(merkleProof, merkleData.root, node);
+  }
+
+  /** 
+   * @dev Claim accumulated rewards for a set of tokens at a given cycle number
+   * @param cycle cycle number
+   * @param index user reward info index in the array of reward info
+   * during merkle tree generation
+   * @param user wallet address of reward beneficiary
+   * @param tokens array of tokens claimable by reward beneficiary
+   * @param cumulativeAmounts cumulative token amounts claimable by reward beneficiary
+   * @param merkleProof merkle proof of claim
+   * @return claimAmounts actual claimed token amounts sent to the reward beneficiary
+   **/
+  function claim(
+    uint256 cycle,
+    uint256 index,
+    address user,
+    IERC20Ext[] calldata tokens,
+    uint256[] calldata cumulativeAmounts,
+    bytes32[] calldata merkleProof
+  ) external override nonReentrant returns (uint256[] memory claimAmounts) {
+    require(cycle == merkleData.cycle, 'incorrect cycle');
```

Comment:
should we just remove this cycle field, and always use `merkleData.cycle` when calling claim function?

Reply:
Mmmm I thought about it, it's useful for when user txns are pending for too long that the next merkle root has been proposed. Will save these users gas instead of validating old merkle proof.

Reply:
ok, I think we can keep it, many users will try to use low gas to claim